sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/Validations.scala

```diff
@@ -0,0 +1,362 @@
+    when(
+      hasEventSourcedEntitySubscription(component) &&
+      methods.exists(hasEventSourcedEntitySubscription)) {
+      // collect offending methods
+      methods.filter(hasEventSourcedEntitySubscription).map { method =>
+        errorMessage(
+          method,
+          "You cannot use @Subscribe.EventSourcedEntity annotation in both methods and class. You can do either one or the other.")
+      }
+    }
+  }
+
+  private def topicSubscriptionValidations(component: Class[_]): Validation = {
+    val methods = component.getMethods.toIndexedSeq
+    when(hasTopicSubscription(component) && methods.exists(hasTopicSubscription)) {
+      // collect offending methods
+      methods.filter(hasTopicSubscription).map { method =>
+        errorMessage(
+          method,
+          "You cannot use @Subscribe.Topic annotation in both methods and class. You can do either one or the other.")
+      }
+    }
+  }
+
+  private def publishStreamIdMustBeFilled(component: Class[_]): Validation = {
+    Option(component.getAnnotation(classOf[Publish.Stream]))
+      .map { ann =>
+        when(ann.id().trim.isEmpty) {
+          Seq("Publish.Stream id can not be an empty string")
+        }
+      }
+      .getOrElse(Valid)
+  }
+
+  private def noSubscriptionMethodWithAcl(component: Class[_]): Validation = {
+
+    val hasSubscriptionAndAcl = (method: Method) => hasAcl(method) && hasSubscription(method)
+
+    val messages =
+      component.getMethods.toIndexedSeq.filter(hasSubscriptionAndAcl).map { method =>
+        errorMessage(
+          method,
+          "Subscription methods are for internal use only and cannot be combined with ACL annotations.")
+      }
+
+    Validation(messages)
+  }
+
+  private def noSubscriptionWithRestAnnotations(component: Class[_]): Validation = {
+
+    val hasSubscriptionAndRest = (method: Method) => hasRestAnnotation(method) && hasSubscription(method)
+
+    val messages =
+      component.getMethods.toIndexedSeq.filter(hasSubscriptionAndRest).map { method =>
+        errorMessage(
+          method,
+          "Methods annotated with Kalix @Subscription annotations can not be annotated with REST annotations.")
+      }
+
+    Validation(messages)
+  }
+
+  private def noRestStreamIn(component: Class[_]): Validation = {
+
+    // this is more for early validation. We don't support stream-in over http,
+    // we block it before deploying anything
+    def isStreamIn(method: Method): Boolean = {
+      val paramWithRequestBody =
+        method.getParameters.collect {
+          case param if param.getAnnotation(classOf[RequestBody]) != null => param
+        }
+      paramWithRequestBody.exists(_.getType == classOf[Flux[_]])
+    }
+
+    val hasRestWithStreamIn = (method: Method) => hasRestAnnotation(method) && isStreamIn(method)
+
+    val messages =
+      component.getMethods.filter(hasRestWithStreamIn).map { method =>
+        errorMessage(method, "Stream in calls are not supported.")
```

Comment:
```suggestion
        errorMessage(method, "@Subscribe.Stream in REST calls is not supported.")
```
Not sure I've got this one right. If not, could you rephrase it?

Reply:
yeah, here the concept is `stream in` calls. It's not about `@Subscribe.Stream`. 

Stream in calls are calls where the client sends an stream as input. This is not supported for REST.